File: backend/app/core/upgrades/curses/damage_reduc_4.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class DamageReduc4(UpgradeImplementation):
    id = 'damage_reduc_4'
    name = 'Weakness 4'
    description = '-1.75% damage.'
    rarity = 'epic'
    type = 'stat_modifier'
    cost = 0

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = -0.0175
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/curses/damage_reduc_4.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const DamageReduc4Def: UpgradeDef = {
  id: "damage_reduc_4",
  name: "Weakness 4",
  description: "-1.75% damage.",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 0,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: -0.0175,
  isMultiplier: true,
  stackable: true,
  maxStacks: 99999,
  curse: true,
}

export class DamageReduc4 extends Upgrade {}
